rust/src/day1.rs

```diff
@@ -23,18 +23,17 @@ fn count_increases(depths: &Vec<u32>) -> u32 {
         .windows(2)
         .filter_map(|w| match w {
             [a, b] => Some((a, b)),
-            _ => None
+            _ => None,
         })
         .fold(0, |n, (d0, d1)| n + u32::from(d1 > d0))
 }
 
 /** @todo Need to understand what a Box is. (Seems to describe a heap allocated element). */
```

Comment:
So, yes, indeed, `Box` is “just” some data on the heap. It’s more magical than that, but that’s usually sufficient to understand why it’s necessary. In your case, you had to use it because you were returning a trait which has many implementations (of different size). This requires putting the data on the heap (because the compiler doesn't know how much space to reserve on the stack). That's the `Box` part. Then, because we don't know what implementation of `Iterator` is in the box, we also don't know where the methods are in memory. So this further requires using dynamic dispatch to reach those methods. That's the `dyn` part.

But Rust is smarter than it was now and you can now tell the compiler that your function returns "some" implementation of `Iterator` by using the [`impl Trait` syntax](https://doc.rust-lang.org/rust-by-example/trait/impl_trait.html). This tells the compiler to generate different versions of the function depending on the value returned.

This can also be used in the argument position of functions, simplifying signatures in other places as well...

Reply:
> In your case, you had to use it because you were returning a trait which has many implementations (of different size)

Thanks for the thorough explanation. Really appreciated. You sort of connect the dots I encountered yesterday.

It all started when I tried to extract the `Iterator` behind a function, and then I have to express the expanded type resulting of the chained methods: 
```rust
Map<FilterMap<Lines<BufReader<File>>, fn(Result<String>) -> Option<String>>, fn(String) -> u32>
```

That's what was the compiler referring to indeed (sized unbound, or impossile to determine size),

Reply:
>  `impl Trait syntax`. This tells the compiler to generate different versions of the function depending on the value returned.

Interesting. I guess that by generating those version of each different usage case, the compiler can infer:
- where the methods are in memory.
- size of the instance.

I'll keep this in my backpocket to come back to it later...